### PR TITLE
[SPARK-36197][SQL] Use PartitionDesc instead of TableDesc for reading hive partitioned tables

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

A hive partition can have different `PartitionDesc`s from `TableDesc` for describing Serde/InputFormatClass/OutputFormatClass, for a hive partitioned table, we shall respect those in `PartitionDesc`.

### Why are the changes needed?

in many cases, that Spark reads hive tables could result in surprise because of this issue.

### Does this PR introduce _any_ user-facing change?

yes, hive partition table that contains different serde/input/output could be recognized by Spark

### How was this patch tested?

new test added

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu thanks

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 1]
```diff
@@ -345,13 +372,26 @@ class HadoopTableReader(
     HadoopTableReader.initializeLocalJobConfFunc(path, tableDesc)(newJobConf)
     val inputFormatClass = tableDesc.getInputFileFormatClass
       .asInstanceOf[Class[newInputClass[Writable, Writable]]]
+    createNewHadoopRDD(inputFormatClass, newJobConf)
+  }
+
+  private def createNewHadoopRDD(partDesc: PartitionDesc, path: String): RDD[Writable] = {
+    val newJobConf = new JobConf(hadoopConf)
+    HadoopTableReader.initializeLocalJobConfFunc(path, partDesc.getTableDesc)(newJobConf)
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:380]: Can we call `createNewHadoopRDD(tableDesc: TableDesc, path: String)` here?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 2]
```diff
@@ -322,7 +332,24 @@ class HadoopTableReader(
...
+    createOldHadoopRDD(inputFormatClass, initializeJobConfFunc)
+  }
 
+  /**
+   * Creates a HadoopRDD based on the broadcasted HiveConf and other job properties that will be
+   * applied locally on each executor.
+   */
+  private def createOldHadoopRDD(partitionDesc: PartitionDesc, path: String): RDD[Writable] = {
+    val initializeJobConfFunc =
+      HadoopTableReader.initializeLocalJobConfFunc(path, partitionDesc.getTableDesc) _
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:344]: Can we call `createOldHadoopRDD(tableDesc: TableDesc, path: String) ` here?

Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:344]: `partitionDesc` is still needed here to `getInputFileFormatClass `

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks all, merged to master and 3.2(as it's also a bug for reading hive tables)

**Pull request closed without merging**